# Pandas data frames

In this notebook, we will start learning about Pandas data frames. To import the pandas library to our notebook, you will first need to download and install the pandas library. You can do this in the terminal by opening up a terminal window "Terminal > New Terminal" and writing the command `python3 -m pip install pandas`.



Then you can import the pandas (and numpy) libraries into this notebook as follows:

In [3]:
import pandas as pd
import numpy as np

To load a .csv data file into our space, we need to use the `read_csv()` function from the pandas library. Make sure that you have saved the `gapminder.csv` file in a `data` subfolder that lives in the same place where this notebook is saved.

Let's load the gapminder dataset:

In [4]:
pd.read_csv('data/gapminder.csv')

country continent  year  lifeExp       pop   gdpPercap
0     Afghanistan      Asia  1952   28.801   8425333  779.445314
1     Afghanistan      Asia  1957   30.332   9240934  820.853030
2     Afghanistan      Asia  1962   31.997  10267083  853.100710
3     Afghanistan      Asia  1967   34.020  11537966  836.197138
4     Afghanistan      Asia  1972   36.088  13079460  739.981106
...           ...       ...   ...      ...       ...         ...
1699     Zimbabwe    Africa  1987   62.351   9216418  706.157306
1700     Zimbabwe    Africa  1992   60.377  10704340  693.420786
1701     Zimbabwe    Africa  1997   46.809  11404948  792.449960
1702     Zimbabwe    Africa  2002   39.989  11926563  672.038623
1703     Zimbabwe    Africa  2007   43.487  12311143  469.709298

[1704 rows x 6 columns]

This just prints out the gapminder dataset, but it doesn't save it. 

To save the dataset so that we can use it in our notebook, we can to assign the results of the `pd.read_csv()` function to a variable called `gapminder`:

In [5]:
gapminder = pd.read_csv('data/gapminder.csv')

To view the dataset we have just loaded, we can type the name of the variable that we saved it in:

In [6]:
gapminder

country continent  year  lifeExp       pop   gdpPercap
0     Afghanistan      Asia  1952   28.801   8425333  779.445314
1     Afghanistan      Asia  1957   30.332   9240934  820.853030
2     Afghanistan      Asia  1962   31.997  10267083  853.100710
3     Afghanistan      Asia  1967   34.020  11537966  836.197138
4     Afghanistan      Asia  1972   36.088  13079460  739.981106
...           ...       ...   ...      ...       ...         ...
1699     Zimbabwe    Africa  1987   62.351   9216418  706.157306
1700     Zimbabwe    Africa  1992   60.377  10704340  693.420786
1701     Zimbabwe    Africa  1997   46.809  11404948  792.449960
1702     Zimbabwe    Africa  2002   39.989  11926563  672.038623
1703     Zimbabwe    Africa  2007   43.487  12311143  469.709298

[1704 rows x 6 columns]

We can then use the same `type()` function that we used in the previous notebook to ask what kind of object the `gapminder` variable is (the answer is a pandas DataFrame):

In [7]:
type(gapminder)

pandas.core.frame.DataFrame

## Extracting information/attributes from DataFrames

In this section, we will learn how to extract attributes from DataFrame objects and how to apply DataFrame-specific "methods" to DataFrames, both using the `.` syntax.

As a reminder, let's print out the `gapminder` DataFrame that we're working with:

In [8]:
gapminder

country continent  year  lifeExp       pop   gdpPercap
0     Afghanistan      Asia  1952   28.801   8425333  779.445314
1     Afghanistan      Asia  1957   30.332   9240934  820.853030
2     Afghanistan      Asia  1962   31.997  10267083  853.100710
3     Afghanistan      Asia  1967   34.020  11537966  836.197138
4     Afghanistan      Asia  1972   36.088  13079460  739.981106
...           ...       ...   ...      ...       ...         ...
1699     Zimbabwe    Africa  1987   62.351   9216418  706.157306
1700     Zimbabwe    Africa  1992   60.377  10704340  693.420786
1701     Zimbabwe    Africa  1997   46.809  11404948  792.449960
1702     Zimbabwe    Africa  2002   39.989  11926563  672.038623
1703     Zimbabwe    Africa  2007   43.487  12311143  469.709298

[1704 rows x 6 columns]

To extract an attribute from an object in Python, we use the `object.attribute` syntax. So if we want to extract the `shape` attribute from the `gapminder` DataFrame object, we can do so as follows:

In [9]:
gapminder.shape

(1704, 6)

This `shape` attribute tells us the number of rows (1704) and the number of columns (6) and is helpful for learning about the size of our data objects

The `head()` function typically prints out the first few rows of a DataFrame. However, `head()` is not a regular function. If `head()` were a regular function, we would be able to apply it like this:

In [10]:
head(gapminder)

NameError: name 'head' is not defined

But this results in an error. This is because `head()` is not a function that can be applied in the regular way. Instead, `head()` is a  special type of function that can only be applied to pandas DataFrame objects. Object-specific functions are called **methods**, and are applied using the `object.method()` syntax rather than the `method(object)` syntax above. 

We can apply the `head()` method to the `gapminder` dataset as follows, which will print out the first 5 rows of the DataFrame:

In [11]:
gapminder.head()

country continent  year  lifeExp       pop   gdpPercap
0  Afghanistan      Asia  1952   28.801   8425333  779.445314
1  Afghanistan      Asia  1957   30.332   9240934  820.853030
2  Afghanistan      Asia  1962   31.997  10267083  853.100710
3  Afghanistan      Asia  1967   34.020  11537966  836.197138
4  Afghanistan      Asia  1972   36.088  13079460  739.981106

You can provide additional arguments to the `head()` inside the parentheses. For example, if you want to print 10 rows instead of 5, you can do so as follows:

In [12]:
gapminder.head(10)

country continent  year  lifeExp       pop   gdpPercap
0  Afghanistan      Asia  1952   28.801   8425333  779.445314
1  Afghanistan      Asia  1957   30.332   9240934  820.853030
2  Afghanistan      Asia  1962   31.997  10267083  853.100710
3  Afghanistan      Asia  1967   34.020  11537966  836.197138
4  Afghanistan      Asia  1972   36.088  13079460  739.981106
5  Afghanistan      Asia  1977   38.438  14880372  786.113360
6  Afghanistan      Asia  1982   39.854  12881816  978.011439
7  Afghanistan      Asia  1987   40.822  13867957  852.395945
8  Afghanistan      Asia  1992   41.674  16317921  649.341395
9  Afghanistan      Asia  1997   41.763  22227415  635.341351

This argument has a name `n`, which you can explicitly specify:

In [13]:
gapminder.head(n=10)

country continent  year  lifeExp       pop   gdpPercap
0  Afghanistan      Asia  1952   28.801   8425333  779.445314
1  Afghanistan      Asia  1957   30.332   9240934  820.853030
2  Afghanistan      Asia  1962   31.997  10267083  853.100710
3  Afghanistan      Asia  1967   34.020  11537966  836.197138
4  Afghanistan      Asia  1972   36.088  13079460  739.981106
5  Afghanistan      Asia  1977   38.438  14880372  786.113360
6  Afghanistan      Asia  1982   39.854  12881816  978.011439
7  Afghanistan      Asia  1987   40.822  13867957  852.395945
8  Afghanistan      Asia  1992   41.674  16317921  649.341395
9  Afghanistan      Asia  1997   41.763  22227415  635.341351

But you don't need to specify the `n=` part of the argument because the `head()` method knows that the first argument is the number of rows to print.

### Exercise

1. The pandas DataFrame has an attribute called `dtypes` that will print out the *type* of each column. Extract the `dtypes` attribute from the `gapminder` DataFrame:

In [14]:
gapminder.dtypes

country       object
continent     object
year           int64
lifeExp      float64
pop            int64
gdpPercap    float64
dtype: object

Note that the "string" type is called `object` in pandas.

2. The pandas DataFrame has a "method" called `select_dtypes()` that will extract just the columns of a certain type from the DataFrame. Use the `select_dtypes()` function to extract the numeric (float and integer) columns of gapminder by providing an argument `include='number'` inside the parentheses of `select_dtypes()`. 

In [15]:
gapminder.select_dtypes(include='number')

year  lifeExp       pop   gdpPercap
0     1952   28.801   8425333  779.445314
1     1957   30.332   9240934  820.853030
2     1962   31.997  10267083  853.100710
3     1967   34.020  11537966  836.197138
4     1972   36.088  13079460  739.981106
...    ...      ...       ...         ...
1699  1987   62.351   9216418  706.157306
1700  1992   60.377  10704340  693.420786
1701  1997   46.809  11404948  792.449960
1702  2002   39.989  11926563  672.038623
1703  2007   43.487  12311143  469.709298

[1704 rows x 4 columns]

In [14]:
# to instead extract the string/object-type columns:
gapminder.select_dtypes(include='object')

country continent
0     Afghanistan      Asia
1     Afghanistan      Asia
2     Afghanistan      Asia
3     Afghanistan      Asia
4     Afghanistan      Asia
...           ...       ...
1699     Zimbabwe    Africa
1700     Zimbabwe    Africa
1701     Zimbabwe    Africa
1702     Zimbabwe    Africa
1703     Zimbabwe    Africa

[1704 rows x 2 columns]

## DataFrame indexing

In this section we will learn about the column and row indexes of the DataFrame. These are essentially the column and row names of the DataFrame.

We will keep working with the gapminder DataFrame that is printed below:

In [15]:
gapminder

country continent  year  lifeExp       pop   gdpPercap
0     Afghanistan      Asia  1952   28.801   8425333  779.445314
1     Afghanistan      Asia  1957   30.332   9240934  820.853030
2     Afghanistan      Asia  1962   31.997  10267083  853.100710
3     Afghanistan      Asia  1967   34.020  11537966  836.197138
4     Afghanistan      Asia  1972   36.088  13079460  739.981106
...           ...       ...   ...      ...       ...         ...
1699     Zimbabwe    Africa  1987   62.351   9216418  706.157306
1700     Zimbabwe    Africa  1992   60.377  10704340  693.420786
1701     Zimbabwe    Africa  1997   46.809  11404948  792.449960
1702     Zimbabwe    Africa  2002   39.989  11926563  672.038623
1703     Zimbabwe    Africa  2007   43.487  12311143  469.709298

[1704 rows x 6 columns]

### The column index

To extract the column index, which corresponds to the column names of the DataFrame, we need to extract the `columns` attribute of the Dataframe object.

In [16]:
gapminder.columns

Index(['country', 'continent', 'year', 'lifeExp', 'pop', 'gdpPercap'], dtype='object')

Notice that the output of the cell above is an "Index" object. If we want to just extract the column values themselves from the index object, we can use the `list()` function to convert the index object to a simpler type of object called a "list" (which is just a collection of values):

In [17]:
list(gapminder.columns)

['country', 'continent', 'year', 'lifeExp', 'pop', 'gdpPercap']

### The row index

The row index can be extracted using the `index` attribute (there is no `rows` attribute):

In [18]:
# row index
gapminder.index

RangeIndex(start=0, stop=1704, step=1)

This time, the output is a `RangeIndex` object, which corresponds to a sequence of integer values with a start value and a stop value with a step size. Since the start is 0, the stop is 1704 (note that the stop is *not* inclusive) and the step size is 1, this RangeIndex corresponds to the integer values 0, 1, 2, 3, ..., 1703. 

To extract the actual integer values from the RangeIndex object, we can convert it to a list using the `list()` function:

In [19]:
list(gapminder.index)

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


You can change the index using the `set_index()` method and providing, for example, a column name as a string.

Let's set the `country` column to be the row index:

In [20]:
# use the set_index() method
gapminder.set_index('country')

continent  year  lifeExp       pop   gdpPercap
country                                                   
Afghanistan      Asia  1952   28.801   8425333  779.445314
Afghanistan      Asia  1957   30.332   9240934  820.853030
Afghanistan      Asia  1962   31.997  10267083  853.100710
Afghanistan      Asia  1967   34.020  11537966  836.197138
Afghanistan      Asia  1972   36.088  13079460  739.981106
...               ...   ...      ...       ...         ...
Zimbabwe       Africa  1987   62.351   9216418  706.157306
Zimbabwe       Africa  1992   60.377  10704340  693.420786
Zimbabwe       Africa  1997   46.809  11404948  792.449960
Zimbabwe       Africa  2002   39.989  11926563  672.038623
Zimbabwe       Africa  2007   43.487  12311143  469.709298

[1704 rows x 5 columns]

Notice that the row index name is on its own line and the original integer row index has disappeared. 

However, notice also that this did not actually modify the `gapminder` object itself. When we print it out below, notice that it is unchanged:

In [21]:
gapminder

country continent  year  lifeExp       pop   gdpPercap
0     Afghanistan      Asia  1952   28.801   8425333  779.445314
1     Afghanistan      Asia  1957   30.332   9240934  820.853030
2     Afghanistan      Asia  1962   31.997  10267083  853.100710
3     Afghanistan      Asia  1967   34.020  11537966  836.197138
4     Afghanistan      Asia  1972   36.088  13079460  739.981106
...           ...       ...   ...      ...       ...         ...
1699     Zimbabwe    Africa  1987   62.351   9216418  706.157306
1700     Zimbabwe    Africa  1992   60.377  10704340  693.420786
1701     Zimbabwe    Africa  1997   46.809  11404948  792.449960
1702     Zimbabwe    Africa  2002   39.989  11926563  672.038623
1703     Zimbabwe    Africa  2007   43.487  12311143  469.709298

[1704 rows x 6 columns]

If we wanted to create a version of the `gapminder` DataFrame with the country column as the index, we need to save it as a new variable (you can instead overwrite the `gapminder` variable with this new version, but this is not recommended because we want to keep an unmodified version of the original dataset accessible in our environment). 

Below, we create a *new* DataFrame corresponding to the version of `gapminder` with the `'country'` column as the row index:

In [22]:
gapminder_country = gapminder.set_index('country')

Notice that the original gapminder dataset is unchanged:

In [23]:
gapminder

country continent  year  lifeExp       pop   gdpPercap
0     Afghanistan      Asia  1952   28.801   8425333  779.445314
1     Afghanistan      Asia  1957   30.332   9240934  820.853030
2     Afghanistan      Asia  1962   31.997  10267083  853.100710
3     Afghanistan      Asia  1967   34.020  11537966  836.197138
4     Afghanistan      Asia  1972   36.088  13079460  739.981106
...           ...       ...   ...      ...       ...         ...
1699     Zimbabwe    Africa  1987   62.351   9216418  706.157306
1700     Zimbabwe    Africa  1992   60.377  10704340  693.420786
1701     Zimbabwe    Africa  1997   46.809  11404948  792.449960
1702     Zimbabwe    Africa  2002   39.989  11926563  672.038623
1703     Zimbabwe    Africa  2007   43.487  12311143  469.709298

[1704 rows x 6 columns]

But that the `gapminder_country` DataFrame has the `country` column as its index. 

In [24]:
gapminder_country

continent  year  lifeExp       pop   gdpPercap
country                                                   
Afghanistan      Asia  1952   28.801   8425333  779.445314
Afghanistan      Asia  1957   30.332   9240934  820.853030
Afghanistan      Asia  1962   31.997  10267083  853.100710
Afghanistan      Asia  1967   34.020  11537966  836.197138
Afghanistan      Asia  1972   36.088  13079460  739.981106
...               ...   ...      ...       ...         ...
Zimbabwe       Africa  1987   62.351   9216418  706.157306
Zimbabwe       Africa  1992   60.377  10704340  693.420786
Zimbabwe       Africa  1997   46.809  11404948  792.449960
Zimbabwe       Africa  2002   39.989  11926563  672.038623
Zimbabwe       Africa  2007   43.487  12311143  469.709298

[1704 rows x 5 columns]